File: 06.o-estilo-de-programacao-funcional/jogodavelha.java
import java.util.Scanner;
public class Main {
	public static void main(String[] args) {
		comecaJogo(0 , 0);
	}
    	private static void comecaJogo(int pontuacaoJogador1, int pontuacaoJogador2) {
                System.out.println("Faça sua escolha: \n(1) jogar\n(2) Ver pontuação dos jogadores\n(3) Sair do jogo");
        	Scanner sc = new Scanner(System.in);
        	int escolha = sc.nextInt();
       	 	switch (escolha){
	            case 1:
		         char[][] board = inicializa();
			 print_board(board);
			 char ganhador = joga(board, 'X');
			 if (ganhador == 'X') pontuacaoJogador1++; else if(ganhador == 'O') pontuacaoJogador2++;
			 comecaJogo(pontuacaoJogador1, pontuacaoJogador2);
		         break;
	            case 2:
	                mostraPontuacao(pontuacaoJogador1, pontuacaoJogador2);
                        comecaJogo(pontuacaoJogador1,pontuacaoJogador2);
                        break;
                    case 3:
                        System.out.println("Espero você de volta");
		}
	}
	private static void mostraPontuacao(int pontuacaoJogador1, int pontuacaoJogador2) {
		System.out.println("Pontuação do jogador 1 é de "+pontuacaoJogador1);
		System.out.println("Pontuação do jogador 2 é de "+pontuacaoJogador2);
	}
        private static char joga(char[][] board, char vezPlayer) {
	    	System.out.println("Vez de " + vezPlayer + " faça sua jogada");
		char[][] novoBoard = marcaPonto(board, vezPlayer);
		print_board(novoBoard);
		if(verificaVitoria(novoBoard)) {
			System.out.println("O player " + vezPlayer + " ganhou!!");
			return vezPlayer;
		}
		else if(verificaEmpate(novoBoard)) {
			System.out.println("Deu empate :p");
			return '_';
		}
		else {
			vezPlayer = (vezPlayer == 'X') ? 'O' : 'X';
			return joga(novoBoard, vezPlayer);
		}
	}
	private static char[][] marcaPonto(char[][] board, char vezPlayer) {
		char[][] novoBoard = repassaPontos(board);
		novoBoard = marcaPontoAux(novoBoard,vezPlayer);
		return novoBoard;
	}
	private static char[][] repassaPontos(char[][] board) {
		char[][] novoBoard = inicializa();
		for(int i = 0 ; i<3 ; i++)
			for(int j = 0 ; j<3 ; j++)
				novoBoard[i][j] = board[i][j];
		return novoBoard;
	}
	private static char[][] marcaPontoAux(char[][]board,char vezPlayer) {
		Scanner sc = new Scanner(System.in);
		System.out.println("Escolha a linha (1-3)");
		int linha = sc.nextInt() - 1;
		System.out.println("Escolha a coluna (1-3)");
		int coluna = sc.nextInt() -1;
		if (linha>=0 && 2>=linha && coluna>=0 && 2>=coluna && board[linha][coluna] ==' ') {
			board[linha][coluna] = vezPlayer;
			return board;
		}
		else {
			System.out.println("Jogada invalida favor repita ela");
			return marcaPonto(board, vezPlayer);
		}
	}
	private static void print_board(char[][] board) {
		for(int i = 0 ; i<3 ; i++) {
			System.out.println("| " + board[i][0] + " | " + board[i][1] + " | " + board[i][2] + " | ");
			System.out.println("|---|---|---|");
		}
	}
    	private static char[][] inicializa() {
    		char[][] board = new char[3][3];
		for(int i = 0 ; i<3 ; i++)
			for(int j = 0 ; j<3 ; j++)
				board[i][j] = ' ';
		return board;
	}
    	private static boolean verificaVitoria(char[][]board) {
    		for(int i = 0 ; i<3 ; i++) {
    			if(board[i][0]==board[i][1] && board[i][0]==board[i][2] && board[i][0]!=' ')
				return true;
    			if(board[0][i]==board[1][i] && board[0][i]==board[2][i] && board[0][i]!=' ')
				return true;
		}
		if((board[0][0]==board[1][1] && board[0][0]==board[2][2] && board[0][0]!=' ') || (board[0][2]==board[1][1] && board[0][2]==board[2][0] && board[0][2]!=' '))
			return true;
		return false;
	}
	private static boolean verificaEmpate(char[][]board) {
        	for(int i = 0 ; i<3 ; i++)
			for(int j = 0 ; j<3 ; j++)
				if(board[i][j] == ' ')
		    			return false;
		return true;
    }	
}

File: 06.o-estilo-de-programacao-funcional/jogodavelha.clj
(ns jogodavelha.core
    (:require [clojure.string :as str]))



(defn inicializa []
  (vec (repeat 3 (vec (repeat 3 \space)))))


(defn printBoard [board]
  (doseq [linha board]
    (do
    (println (str "| " (nth linha 0) " | " (nth linha 1) " | " (nth linha 2) " |"))
    (println "|---|---|---|"))))


(defn mostraPontuacao [score1 score2]
  (println (str "Pontuação do jogador 1 é de : " score1))
  (println (str "Pontuação do jogador 2 é de : " score2))) 


(defn marcaPonto [board vez-player]
  (println "Escolha a linha (1-3)")
  (let [linha (- (read) 1)]
    (println "Escolha a coluna (1-3)")
    (let [coluna (- (read) 1)]
      (if (and (<= 0 linha) (<= linha 2)  (<= 0 coluna) (<= coluna 2) (= (get-in board [linha coluna]) \space))
        (assoc-in board [linha coluna] vez-player)
        (do
          (println "Jogada inválida, favor repetir.")
          (marcaPonto board vez-player))))))



(defn verificaVitoria [board]
  (if (or (and (= (get-in board [0 0]) (get-in board [1 1]) (get-in board [2 2])) (not= (get-in board [0 0]) \space))
          (and (= (get-in board [0 2]) (get-in board [1 1]) (get-in board [2 0])) (not= (get-in board [0 2]) \space)))
    true
   (loop [i 0]
     (if (< i 3)
      (let [linha (get board i)]
        (if (or (and (= (get linha 0) (get linha 1) (get linha 2)) (not= (get linha 0) \space))
                (and (= (get-in board [0 i]) (get-in board [1 i]) (get-in board [2 i])) (not= (get-in board [0 i]) \space)))
          true
          (recur (inc i))))
      false))))


(defn not-space? [caracter]
    (not= \space caracter))


(defn verificaEmpate [board]
    (every? (fn [linha] (every? not-space? linha)) board))


(defn joga [board vezPlayer]
  (println (str "Vez de " vezPlayer " faça sua jogada"))
  (let [novoBoard (marcaPonto board vezPlayer)]
    (printBoard novoBoard)
    (cond
      (verificaVitoria novoBoard) (do (println (str "O player " vezPlayer " ganhou!!")) vezPlayer)
      (verificaEmpate novoBoard) (do (println "Deu empate :p") \_)
      :else (joga novoBoard (if (= vezPlayer \X) \O \X)))))

(defn limpaBuffer []
    (let [apenasLimpaBuffer (read-line)]))


(defn comecaJogo [score1 score2]
  (println "Faça sua escolha: \n(1) jogar\n(2) Ver pontuação dos jogadores\n(3) Sair do jogo")
  (let [escolha (read-line)]
    (flush)
    (case escolha
      "1" (do
            (let [board (inicializa)]
              (printBoard board)
              (let [ganhador (joga board \X)]
                (cond
                  (= ganhador \X) (do (limpaBuffer) (comecaJogo (inc score1) score2))
                  (= ganhador \O) (do (limpaBuffer) (comecaJogo score1 (inc score2)))
                  :else           (comecaJogo score1 score2)))))
      "2" (do
            (mostraPontuacao score1 score2)
            (recur score1 score2))
      "3" (println "Espero você de volta")
      (do
        (println "Escolha inválida, tente novamente.")
        (comecaJogo score1 score2)))))


(defn main[]
  (comecaJogo 0 0))



(main)
